rosbag2_compression/src/rosbag2_compression/zstd_decompressor.cpp

```diff
@@ -0,0 +1,242 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <cstdio>
+#include <sstream>
+#include <string>
+#include <vector>
+
+#include "rcpputils/filesystem_helper.hpp"
+
+#include "rosbag2_compression/zstd_decompressor.hpp"
+
+#include "rosbag2_storage/filesystem_helper.hpp"
+
+#include "logging.hpp"
+
+namespace
+{
+
+// String constant used to identify ZstdDecompressor.
+constexpr const char kDecompressionIdentifier[] = "zstd";
+
+/**
+ * Open a file using the C API.
+ * This function calls OS-specific implementation of fopen.
+ *
+ * \param uri is the path to the file
+ * \param read_mode is the read mode string accepted by fopen.
+ * \return the FILE pointer or nullptr if the file was not opened.
+ */
+FILE * open_file(const std::string & uri, const std::string & read_mode)
+{
+  FILE * fp{nullptr};
+#ifdef _WIN32
+  fopen_s(&fp, uri.c_str(), read_mode.c_str());
+#else
+  fp = std::fopen(uri.c_str(), read_mode.c_str());
+#endif
+  return fp;
+}
+
+/**
+ * Read a file from the supplied uri into a vector.
+ *
+ * \param uri is the path to the file.
+ * \return the contents of the buffer as a vector.
+ */
+std::vector<uint8_t> get_input_buffer(const std::string & uri)
+{
+  // Get the file size
+  const auto compressed_buffer_length = rosbag2_storage::FilesystemHelper::get_file_size(uri);
```

Comment:
I think we should react here in case the returned filesize is 0 (aka the file is non-existing)